src/Control/RequestHandler.php

```diff
@@ -546,4 +547,114 @@ public function setRequest($request)
         $this->request = $request;
         return $this;
     }
+
+    /**
+     * Returns a link to this controller. Overload with your own Link rules if they exist.
+     *
+     * @param string $action Optional action
+     * @return string
+     */
+    public function Link($action = null)
+    {
+        // Check configured url_segment
+        $url = $this->config()->get('url_segment');
+        if ($url) {
+            return Controller::join_links($url, $action, '/');
+        }
+
+        // no link defined by default
+        trigger_error(
+            'Request handler '.get_class($this). ' does not have a url_segment defined. '.
+            'Relying on this link may be an application error',
+            E_USER_WARNING
+        );
+        return null;
+    }
+
+    /**
+     * Redirect to the given URL.
+     *
+     * @param string $url
+     * @param int $code
+     * @return HTTPResponse
+     */
+    public function redirect($url, $code = 302)
+    {
+        $url = Director::absoluteURL($url);
+        $response = new HTTPResponse();
+        return $response->redirect($url, $code);
+    }
+
+    /**
+     * Safely get the value of the BackURL param, if provided via querystring / posted var
+     *
+     * @return string
+     */
+    public function getBackURL()
+    {
+        $request = $this->getRequest();
+        if (!$request) {
+            return null;
+        }
+        $backURL = $request->requestVar('BackURL');
+        // Fall back to X-Backurl header
+        if (!$backURL && $request->isAjax() && $request->getHeader('X-Backurl')) {
+            $backURL = $request->getHeader('X-Backurl');
+        }
+        if (!$backURL) {
+            return null;
+        }
+        if (Director::is_site_url($backURL)) {
+            return $backURL;
+        }
+        return null;
+    }
+
+    /**
+     * Get referrer to redirect back to and safely validates it
+     *
+     * @internal called from {@see Form::getValidationErrorResponse}
+     * @return string|null
+     */
+    protected function getReferer()
+    {
+        $request = $this->getRequest();
+        if (!$request) {
+            return null;
+        }
+        $pageURL = $request->getHeader('Referer');
+        if (!$pageURL) {
+            return null;
+        }
+        if (Director::is_site_url($pageURL)) {
```

Comment:
Well, I can call it `getInternalReferer`, but the intent of this method is to get only a referer that's safe to redirect to.

I could add a flag to it to allow external urls to be allowed?